pkg/experiment/ingester/memdb/profile_index.go

```diff
@@ -0,0 +1,176 @@
+package memdb
+
+import (
+	"context"
+	memindex "github.com/grafana/pyroscope/pkg/experiment/ingester/memdb/index"
+	phlaremodel "github.com/grafana/pyroscope/pkg/model"
+	schemav1 "github.com/grafana/pyroscope/pkg/phlaredb/schemas/v1"
+	"github.com/grafana/pyroscope/pkg/phlaredb/tsdb"
+	"github.com/grafana/pyroscope/pkg/phlaredb/tsdb/index"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/storage"
+	"go.uber.org/atomic"
+	"sort"
+	"sync"
+)
+
+type profileSeries struct {
+	lbs              phlaremodel.Labels
+	fp               model.Fingerprint
+	minTime, maxTime int64
+	profiles         []*schemav1.InMemoryProfile
+}
+
+type profilesIndex struct {
+	ix              *tsdb.BitPrefixInvertedIndex
+	profilesPerFP   map[model.Fingerprint]*profileSeries
+	mutex           sync.RWMutex
+	rowGroupsOnDisk int
+
+	metrics     *HeadMetrics
+	totalSeries *atomic.Int64
+}
+
+func newProfileIndex(metrics *HeadMetrics) *profilesIndex {
+	ix, err := tsdb.NewBitPrefixWithShards(32)
+	if err != nil {
+		panic(err)
+	}
+	return &profilesIndex{
+		ix:            ix,
+		profilesPerFP: make(map[model.Fingerprint]*profileSeries),
+		metrics:       metrics,
+		totalSeries:   atomic.NewInt64(0),
+	}
+}
+
+// Add a new set of profile to the index.
+// The seriesRef are expected to match the profile labels passed in.
+func (pi *profilesIndex) Add(ps *schemav1.InMemoryProfile, lbs phlaremodel.Labels, profileName string) {
+	pi.mutex.Lock()
+	defer pi.mutex.Unlock()
+	profiles, ok := pi.profilesPerFP[ps.SeriesFingerprint]
+	if !ok {
+		lbs := pi.ix.Add(lbs, ps.SeriesFingerprint)
+		profiles = &profileSeries{
+			lbs:     lbs,
+			fp:      ps.SeriesFingerprint,
+			minTime: ps.TimeNanos,
+			maxTime: ps.TimeNanos,
+		}
+		pi.profilesPerFP[ps.SeriesFingerprint] = profiles
+		//pi.metrics.series.Set(float64(pi.totalSeries.Inc())) // todo how did it work?
+		pi.totalSeries.Inc()
+		pi.metrics.seriesCreated.WithLabelValues(profileName).Inc()
+	}
+
+	// profile is latest in this series, use a shortcut
+	if ps.TimeNanos > profiles.maxTime {
+		// update max timeNanos
+		profiles.maxTime = ps.TimeNanos
+
+		// add profile to in memory slice
+		profiles.profiles = append(profiles.profiles, ps)
+	} else {
+		// use binary search to find position
+		i := sort.Search(len(profiles.profiles), func(i int) bool {
+			return profiles.profiles[i].TimeNanos > ps.TimeNanos
+		})
+
+		// insert into slice at correct position
+		profiles.profiles = append(profiles.profiles, &schemav1.InMemoryProfile{})
+		copy(profiles.profiles[i+1:], profiles.profiles[i:])
+		profiles.profiles[i] = ps
```

Comment:
We probably should handle a case when the same profile is written twice (e.g., due to client retry). I thought that we handle this case properly, but it does not seem so

We don't have to implement it right now – just want to pay your attention

Reply:
Yeah. I could not find de duplication in the old implementation as well. 

Creatad an issue to not forget https://github.com/orgs/grafana/projects/425/views/9?pane=issue&itemId=76384617